File: exponentiation_by_squaring/exponentiationBySquaring.js
function exponentiationBySquaring (base, power) {
	/*
	Performs Exponentiation By Squaring.
	:param base: Base of expression.
	:param power: Power of expression.
	:return: Returns (base ^ power).
	*/
	if (power < 0) { // When Power is Negative
		return exponentiationBySquaring(1 / base, -power);
	} else if (power === 0) { // Power = 0 (Base Case 1)
		return 1;
	} else if (power === 1) { // Power = 1 (Base Case 2)
		return base;
	} else if (power % 2 === 0) { // When Power is Even
		return exponentiationBySquaring(base * base, power >> 1);
	} else { // When Power is Odd
		return base * exponentiationBySquaring(base * base, (power - 1) >> 1);
	}
}

function main () {
	let base = 3;
	let power = 6;
	let res = exponentiationBySquaring(base, power);
	console.log('%d\n', res);
}

main();

File: exponentiation_by_squaring/exponentiationBySquaring.cpp
#include <iostream>
using namespace std;
int exponentiationBySquaring(int base, int power) {
    if(power < 0) 
        return exponentiationBySquaring(1/base, -power);
    else if(power == 0) {
        return 1; //Because anythign raised to the power 0 is 1.
    }
    else if(power == 1)
        return base; // Return Base number if power is 1.
    else if(power %2 == 0)
        return exponentiationBySquaring(base * base , power/2 ) ;
    else  
        return base * exponentiationBySquaring(base* base, (power-1)/2); 
}
int main()
{
    cout<<exponentiationBySquaring(12,2);
    return 0;
}
